frontend/model/captureLogs.js

```diff
@@ -0,0 +1,247 @@
+  // here in a roundabout way by getting the function to which they're mapped.
+  // The reason this works is because the entire `sbp` domain is blacklisted
+  // from being logged in main.js.
+  sbp('sbp/selectors/fn', 'okTurtles.events/emit')(CAPTURED_LOGS, lastEntry)
+}
+
+function verifyLogsSize () {
+  let lastEntriesCount = null
+  let markers = null
+
+  // Delete nth oldest logs recursively. This scenario can happen when the
+  // entries limit is changed. Example: There are 400 logs and the limit is 500.
+  // We change the limit to 100 and the marker nth to 25. 325 logs need to be deleted.
+  // We do it recursively in chunks of 25 until there's only 75 logs again.
+  while (entriesCount >= ENTRIES_LIMIT) {
+    if (entriesCount === lastEntriesCount) {
+      // There are too many entriesCount, however, for some unknown reason
+      // we weren't able to remove them. Call clearLogs to reset everything.
+      clearLogs()
+      console.error('verifyLogsSize(): unable to delete oldest logs, had to clear them!')
+      return
+    }
+    lastEntriesCount = entriesCount
+
+    markers = markers || getMarkers()
+    let toDelete = ENTRIES_MARKER_NTH
+    let oldestEntry = markers.shift() // the oldest marker
+
+    do {
+      const log = JSON.parse(giLSget(oldestEntry)) || {}
+      giLSremove(oldestEntry)
+      entriesCount--
+      toDelete--
+      oldestEntry = log.prev
+    } while (toDelete && oldestEntry)
+
+    giLSset('markers', JSON.stringify(markers))
+    giLSset('count', entriesCount)
+  }
+}
+
+function setConfig () {
+  giLSset('limit', ENTRIES_LIMIT)
+  giLSset('markerNth', ENTRIES_MARKER_NTH)
+}
+
+function verifyLogsConfig () {
+  lastEntry = giLSget('lastEntry')
+  entriesCount = +giLSget('count') || 0
+
+  // if one of these is empty, it means it's a fresh start
+  if (!entriesCount || !lastEntry) {
+    clearLogs() // make sure we don't leave any lost log behind
+    setConfig()
+    return
+  }
+
+  // If ENTRIES_LIMIT or ENTRIES_MARKER_NTH are changed in a release
+  // we need to recalculate markers and verify if it reached the size limit
+  const storedLimit = +giLSget('limit')
+  const storedMarkerNth = +giLSget('markerNth')
+  setConfig()
+
+  if (storedMarkerNth !== ENTRIES_MARKER_NTH) {
+    // recalculate markers based on new nth rule.
+    const newMarkers = []
+    let entryIndex = entriesCount
+    let prevEntry = lastEntry
+    do {
+      const log = JSON.parse(giLSget(`${prevEntry}`))
+      if (!log) break
+      if (entryIndex % ENTRIES_MARKER_NTH === 0) {
+        newMarkers.push(prevEntry)
+      }
+      entryIndex--
+      prevEntry = log.prev
+    } while (prevEntry)
+    // reverse new markers to be in chronological order, and then store them.
+    giLSset('markers', JSON.stringify(newMarkers.reverse()))
+    giLSset('markerNth', ENTRIES_MARKER_NTH)
+  }
+
+  if (storedLimit !== ENTRIES_LIMIT) {
+    verifyLogsSize()
+  }
+}
```

Comment:
This all is unnecessarily long and complicated. Please replace with:

```js
function verifyLogsConfigs () {
  lastEntry = giLSget('lastEntry')
  entriesCount = +giLSget('count') || 0

  // NOTE: consider calculating the size of entriesCount manually
  //       here (in a loop) if we discover that value gets out of sync.

  // If ENTRIES_LIMIT or ENTRIES_MARKER_NTH are changed in a release
  // we need to recalculate markers and verify if it reached the size limit.
  const storedLimit = +giLSget('limit')
  const storedMarkerNth = +giLSget('markerNth')

  giLSset('count', entriesCount)
  giLSset('limit', ENTRIES_LIMIT)
  giLSset('markerNth', ENTRIES_MARKER_NTH)

  if (storedMarkerNth !== ENTRIES_MARKER_NTH) {
    // recalculate markers based on new nth rule.
    const newMarkers = []
    let entryIndex = entriesCount
    let curEntry = lastEntry
    while (curEntry) {
      const log = JSON.parse(giLSget(curEntry))
      if (!log) break
      if (entryIndex % ENTRIES_MARKER_NTH === 0) {
        newMarkers.push(prevEntry)
      }
      entryIndex--
      curEntry = log.prev
    }
    // reverse new markers to be in chronological order, and then store them.
    giLSset('markers', JSON.stringify(newMarkers.reverse()))
  }
  // will delete logs if necessary due to any changes in ENTRIES_LIMIT
  verifyLogsSize()
}
```

This does the same thing, is many lines shorter, and easier to think about. It will also handle correctly in the `while` loop thet case where `curEntry` is  null.

Reply:
>  is many lines shorter

May I ask, just to double-check, what are the *many* lines shorter? The only diff is the `while...` instead of `do... while` and the last if removal. (changed both). All the rest seems the same.

Let me reinforce that the first if `// if one of these is empty, it means it's a fresh start` is necessary to prevent any possible memory leak. Just a double-check sanity.

Reply:
>  what are the many lines shorter?

It's 9 line shorter (12 lines shorter if you ignore the new comment I added).

> The only diff is the while... instead of do... while and the last if removal. (changed both). All the rest seems the same.

It's not the same. It removes the confusing branching, the unnecessary `setConfig` function, and makes the entire function simpler.

The simplest solution is the one we want to go with.

> Let me reinforce that the first if // if one of these is empty, it means it's a fresh start is necessary to prevent any possible memory leak. Just a double-check sanity.

AFAICT the code above does the same thing, but in a simpler way.

Reply:
> AFAICT the code above does the same thing, but in a simpler way.

Not, it doesn't call `clearLogs()` if `entriesCount` or `lastEntry` don't exist.

Reply:
> It doesn't call clearLogs() if entriesCount or lastEntry don't exist. We need that.

Why do we need that?

Reply:
Sometimes when we log out the LS is cleaned but a new log is stored at the exact same moment and it isn't removed. It's the "[sbp] logout" log. So, when we login again, we'll detect that and remove it.

Another reason is to prevent any bug caused by a previous "accidental" modification to the LS.

Reply:
OK, The way this is currently written just does not make sense to me. :-\

It makes no sense why we should ever call `clearLogs()`  on login. If you're not OK with calculating `entriesCount` because of "accidental modification to LS", then using your same reasoning we should be OK with not looping through the entire LS on login to check for whoknowswhat.

Second, I am pretty sure we need to call this function (`verifyLogsConfigs()`) from `clearLogs()` to reset all of the LS parameters (see my other comment). That cannot be done if `verifyLogsConfigs` calls `clearLogs`.

Reply:
Okay, condition removed. If we found a proper major problem in the future, we'll fix it.